[Bug]: ElasticsearchStore: Exception when using Hybrid Search

### Bug Description

When Hybrid mode is on, we use RRF to normalise the scoring across both BM25 and vector search. When RRF is used, Elasticsearch does not return the score for each hit. 

In the PR https://github.com/jerryjliu/llama_index/pull/7792 we normalise the similarities scores of the top_k scores. This causes an exception as scores aren't present in the response, only rank.

### Version

main

### Steps to Reproduce

Run the elasticsearchstore integration tests with trial licence.  

### Relevant Logs/Tracbacks

```shell
use_async = False

    @pytest.mark.skipif(
        elasticsearch_not_available,
        basic_license,
        reason="elasticsearch is not available or license is basic",
    )
    @pytest.mark.asyncio
    @pytest.mark.parametrize("use_async", [True, False])
    async def test_add_to_es_and_hybrid_query(
        es_store: ElasticsearchStore,
        node_embeddings: List[TextNode],
        use_async: bool,
    ) -> None:
        if use_async:
            await es_store.async_add(node_embeddings)
            res = await es_store.aquery(
                VectorStoreQuery(
                    query_str="lorem",
                    query_embedding=[1.0, 0.0, 0.0],
                    mode=VectorStoreQueryMode.HYBRID,
                    similarity_top_k=1,
                )
            )
        else:
            es_store.add(node_embeddings)
>           res = es_store.query(
                VectorStoreQuery(
                    query_str="lorem",
                    query_embedding=[1.0, 0.0, 0.0],
                    mode=VectorStoreQueryMode.HYBRID,
                    similarity_top_k=1,
                )
            )

test_elasticsearch.py:254: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../llama_index/vector_stores/elasticsearch.py:449: in query
    return asyncio.get_event_loop().run_until_complete(
../../.venv/lib/python3.10/site-packages/nest_asyncio.py:99: in run_until_complete
    return f.result()
/Users/joe/.pyenv/versions/3.10.3/lib/python3.10/asyncio/futures.py:201: in result
    raise self._exception
/Users/joe/.pyenv/versions/3.10.3/lib/python3.10/asyncio/tasks.py:232: in __step
    result = coro.send(None)
../../llama_index/vector_stores/elasticsearch.py:567: in aquery
    similarities=_to_llama_similarities(top_k_scores),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scores = [None]

    def _to_llama_similarities(scores: List[float]) -> List[float]:
        scores_to_norm: np.ndarray = np.array(scores)
>       return np.exp(scores_to_norm - np.max(scores_to_norm)).tolist()
E       TypeError: unsupported operand type(s) for -: 'NoneType' and 'NoneType'

../../llama_index/vector_stores/elasticsearch.py:133: TypeError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 # ChangeLog
 
+## Unreleased
+
+### Bug Fixes / Nits
+- Fix elasticsearch hybrid scoring (#7852)
+
 ## [0.8.35] - 2023-09-27
 
 ### Bug Fixes / Nits

--- a/llama_index/vector_stores/elasticsearch.py
+++ b/llama_index/vector_stores/elasticsearch.py
@@ -530,7 +530,8 @@ class ElasticsearchStore(VectorStore):
         top_k_nodes = []
         top_k_ids = []
         top_k_scores = []
-        for hit in response["hits"]["hits"]:
+        hits = response["hits"]["hits"]
+        for hit in hits:
             source = hit["_source"]
             metadata = source.get("metadata", None)
             text = source.get(self.text_field, None)
@@ -562,7 +563,12 @@ class ElasticsearchStore(VectorStore):
                 )
             top_k_nodes.append(node)
             top_k_ids.append(node_id)
-            top_k_scores.append(hit["_score"])
+            top_k_scores.append(hit.get("_rank", hit["_score"]))
+
+        if query.mode == VectorStoreQueryMode.HYBRID:
+            total_rank = sum(top_k_scores)
+            top_k_scores = [total_rank - rank / total_rank for rank in top_k_scores]
+
         return VectorStoreQueryResult(
             nodes=top_k_nodes,
             ids=top_k_ids,

--- a/tests/vector_stores/test_elasticsearch.py
+++ b/tests/vector_stores/test_elasticsearch.py
@@ -38,8 +38,8 @@ try:
 
     es_license = es_client.license.get()
     # TODO: Fix hybrid search tests
-    # basic_license: bool = "basic" == es_license["license"]["type"]
-    basic_license = True
+    basic_license: bool = "basic" == es_license["license"]["type"]
+    # basic_license = True
 except (ImportError, Exception):
     elasticsearch_not_available = True
     basic_license = True
@@ -408,6 +408,30 @@ async def test_add_to_es_and_text_query_ranked(
     )
     await check_top_match(
         es_store, node_embeddings, use_async, query_get_2_first, node2, node1
+    )
+
+
+@pytest.mark.skipif(
+    elasticsearch_not_available, reason="elasticsearch is not available"
+)
+@pytest.mark.asyncio
+@pytest.mark.parametrize("use_async", [True, False])
+async def test_add_to_es_and_text_query_ranked_hybrid(
+    es_store: ElasticsearchStore,
+    node_embeddings: List[TextNode],
+    use_async: bool,
+) -> None:
+    node1 = "f658de3b-8cef-4d1c-8bed-9a263c907251"
+    node2 = "0b31ae71-b797-4e88-8495-031371a7752e"
+
+    query_get_1_first = VectorStoreQuery(
+        query_str="I was",
+        query_embedding=[0.0, 0.0, 0.5],
+        mode=VectorStoreQueryMode.HYBRID,
+        similarity_top_k=2,
+    )
+    await check_top_match(
+        es_store, node_embeddings, use_async, query_get_1_first, node1, node2
     )
 
 

--- a/tests/vector_stores/test_elasticsearch.py
+++ b/tests/vector_stores/test_elasticsearch.py
@@ -37,9 +37,7 @@ try:
     elasticsearch_not_available = False
 
     es_license = es_client.license.get()
-    # TODO: Fix hybrid search tests
     basic_license: bool = "basic" == es_license["license"]["type"]
-    # basic_license = True
 except (ImportError, Exception):
     elasticsearch_not_available = True
     basic_license = True

